core/include/scipp/core/multi_index.h

```diff
@@ -30,276 +47,332 @@ auto get_slice_dim(const T &param, const Ts &... params) {
-      strides[data] = m_stride[data][0];
-    }
-    return strides;
+    return scipp::span<const scipp::index>(&stride(0, 0), N);
   }
 
-  [[nodiscard]] constexpr scipp::index inner_distance_to_end() const noexcept {
-    return m_shape[0] - m_coord[0];
+  [[nodiscard]] scipp::index inner_distance_to_end() const noexcept {
+    return shape(0) - coord(0);
   }
 
-  [[nodiscard]] constexpr scipp::index
+  [[nodiscard]] scipp::index
   inner_distance_to(const MultiIndex &other) const noexcept {
-    return other.m_coord[0] - m_coord[0];
+    return other.coord(0) - coord(0);
   }
 
   /// Set the absolute index. In the special case of iteration with bins,
   /// this sets the *index of the bin* and NOT the full index within the
   /// iterated data.
-  constexpr void set_index(const scipp::index offset) noexcept {
-    auto remainder{offset};
-    // This loop is until NDIM_MAX inclusive to handle the end index
-    for (scipp::index d = 0; d < NDIM_MAX + 1; ++d) {
-      if (has_bins() && d < m_ndim_nested) {
-        m_coord[d] = 0;
-      } else {
-        if (m_shape[d] == 0) {
-          m_coord[d] = remainder; // this serves as the end index
-          break;
-        }
-        m_coord[d] = remainder % m_shape[d];
-        remainder /= m_shape[d];
+  void set_index(const scipp::index index) noexcept {
+    if (has_bins()) {
+      set_bins_index(index);
+    } else {
+      extract_indices(index, shape_it(), shape_it(m_inner_ndim), coord_it());
+      for (scipp::index data = 0; data < N; ++data) {
+        m_data_index[data] = flat_index(data, 0, m_inner_ndim);
       }
     }
-    for (scipp::index data = 0; data < N; ++data) {
-      m_data_index[data] = flat_index_from_strides(
-          m_stride[data].begin(), m_stride[data].begin() + m_ndim,
-          m_coord.begin());
-    }
+  }
+
+  void set_to_end() noexcept {
     if (has_bins()) {
+      set_to_end_bin();
+    } else {
+      if (m_inner_ndim == 0) {
+        coord(0) = 1;
+      } else {
+        std::fill(coord_it(0), coord_it(m_inner_ndim - 1), 0);
+        coord(m_inner_ndim - 1) = shape(m_inner_ndim - 1);
+      }
       for (scipp::index data = 0; data < N; ++data) {
-        m_bin[data].m_bin_index = flat_index_from_strides(
-            m_stride[data].begin() + m_ndim_nested,
-            m_stride[data].begin() + m_ndim, m_coord.begin() + m_ndim_nested);
-        load_bin_params(data);
+        m_data_index[data] = flat_index(data, 0, m_inner_ndim);
       }
-      if (m_shape[m_nested_dim_index] == 0 && offset != m_end_sentinel)
-        seek_bin();
     }
   }
 
-  constexpr auto get() const noexcept { return m_data_index; }
+  [[nodiscard]] constexpr auto get() const noexcept { return m_data_index; }
 
-  constexpr bool operator==(const MultiIndex &other) const noexcept {
-    return m_coord == other.m_coord;
+  bool operator==(const MultiIndex &other) const noexcept {
+    // Assuming the number dimensions match to make the check cheaper.
```

Comment:
I think this is a given, just like you cannot compare iterators of different std containers?

Reply:
I guess, it makes no sense to compare MultiIndexes that belong to different combinations of Variables anyway. Do you want me to remove the comment?

Reply:
I don't really care, I think. Maybe?